examples/cloud-operations/unmanaged-instances-healthcheck/function/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthchecker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	pubsub "cloud.google.com/go/pubsub"
+	compute "google.golang.org/api/compute/v1"
+)
+
+// HealthChecker
+type HealthChecker struct {
+	ctx            context.Context
+	config         *Configuration
+	computeService *compute.Service
+	pubsubTopic    *pubsub.Topic
+}
+
+// NewHealthChecker creates new HealthChecker
+func NewHealthChecker(context context.Context, config *Configuration, computeService *compute.Service, pubsubTopic *pubsub.Topic) *HealthChecker {
+	return &HealthChecker{
+		ctx:            context,
+		config:         config,
+		computeService: computeService,
+		pubsubTopic:    pubsubTopic,
+	}
+}
+
+// checkTcpPort returns nil when TCP port is open or error when connection failed.
+func checkTcpPort(ip string, port int, timeout time.Duration) error {
+	target := fmt.Sprintf("%s:%d", ip, port)
+
+	conn, err := net.DialTimeout("tcp", target, timeout)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	return nil
+}
+
+// Start - Start Healthchecker
+func (hc *HealthChecker) Start() error {
+
+	wg := sync.WaitGroup{}
+
+	// Limit max parraller goroutines to prevent exosting socket connections
+	guard := make(chan struct{}, hc.config.max_parallelism)
```

Comment:
Just thinking out loud here, but I wonder if `GOMAXPROCS` has an effect here (I think probably not as most time is spent blocking on a syscall).